Alex and Happy Part 2/article.md

```diff
@@ -0,0 +1,1297 @@
+```sml
+let the_answer : int =
+  let a = 20 in
+  let b = 1 in
+  let c = 2 in
+  a * c + b * c
+
+let main (unit : ()) : () =
+  print ("The answer is: " + the_answer)
+```
+
+Some of the features that this tutorial will support are:
+* Variables, constants, function declarations, function applications, local bindings, and operators.
+* Some simple types, as well as arrow types (functions).
+* Conditional expressions: if-then-else and if-then.
+* Block comments (added in part 1).
+* The reader will be invited in the proposed exercises to extend the language with patterns, list access, and pattern matching. The exercises in the first part also introduced rational numbers and line comments.
+
+## Comparison with parser combinators
+
+<!--
+* [X] refer to parser combinators article
+* [X] talk about the pros and cons
+-->
+
+In a previous tutorial, we discussed implementing and using [Parser Combinators in Haskell](https://serokell.io/blog/parser-combinators-in-haskell).
+In the article, we discussed some pros and cons of LL parsers compared to LR parsers, which we will reiterate and elaborate on.
+
+<!--
+TODO: not sure how correct is the theory here, a more in-depth review here would be nice
+-->
+
+Haskell libraries such as Megaparsec are typically exposed as a LL(1) or LL(∞) parser.
+LL here means **L**eft-to-right parsing, **L**eftmost derivation, as they scan strings from left to right and build a tree deriving them from the left (see below).
+
+The number between the parentheses indicates how many look-ahead tokens the parser can see.
+A parser with a look-ahead of `k` can make decisions by looking `k` tokens (or characters) ahead.
+A look-ahead of `k=1` is often better for better performance, and it's enough to parse most grammars in programming languages.
+
+On the other hand, Happy is an LR(1)\* parser, where the R means **R**ightmost derivation in reverse.
+
+\* Happy is actually an LALR(1) (Look-Ahead LR) parser, which is a more performant albeit less powerful variant of an LR(1) parser; see [section 8.4.1](https://monlih.github.io/happy-docs/#_sec_conflict_tips) of the Happy User Guide.
+Happy also supports GLR (Generalized LR) parsing, which this tutorial won't cover, but you can find more about it in the [section 3](https://monlih.github.io/happy-docs/#_sec_glr) of the Happy User Guide.
+
+As an example, suppose we have the expression `1 + 2 + 3`, which is parsed as the following tree:
+
+```text
+                +
+              ↙   ↘
+            +       3
+          ↙   ↘
+        1       2
+```
+
+It means that in an expression such as `1 + 2 + 3`, a LL parser would expand the tree visiting the leftmost nodes first, visiting the tokens in this order:
+
+<pre>
+<u>exp</u>
+<b><u>exp</u> + exp</b>
+<b><u>exp</u> + exp</b> + exp
+<b>1</b> + <u>exp</u> + exp
+1 + <b>2</b> + <u>exp</u>
+1 + 2 + <b>3</b>
+</pre>
+
+n.b.: **Bold** means the result of the derivation, and <u>underline</u> means the derivation that will be expanded next.
```

Comment:
I don't understand this diagram and underline doesn't work in GitHub preview for some reason

Reply:
As per @lierdakil's suggestion, I've defined the grammar for it, I hope it will be a bit clearer. But if it isn't, tell me and I will change it.

I replaced the underline with italic, but now the bold doesn't work. :facepalm: I will find a solution...